pkg/registry/core/pod/strategy.go

```
@@ -292,10 +295,9 @@ var ResizeStrategy = podResizeStrategy{
 
 // dropNonResizeUpdates discards all changes except for pod.Spec.Containers[*].Resources, pod.Spec.InitContainers[*].Resources, ResizePolicy and certain metadata
 func dropNonResizeUpdates(newPod, oldPod *api.Pod) *api.Pod {
-	pod := dropPodUpdates(newPod, oldPod)
-
 	// Containers are not allowed to be re-ordered, but in case they were,
 	// we don't want to corrupt them here. It will get caught in validation.
```

will things like reordering or name changing or duplicate names actually get caught in validation if this function is stomping containers / initContainers with this merged list prior to validation?

---

I *think* if newPod.Spec.Containers adds new named containers, the loop will silently fail to find the container in the old pod and ignore the change

if newPod.Spec.Containers reorders containers or switches names around, the loop will silently apply any resource changes to the correspondingly named containers in the old pod and then overwrite newPod.Spec.Containers with the result

I agree we *shouldn't* accept resize requests that attempt to add / remove / reorder / renaming containers, but I think we currently can while silently throwing away those attempts

---

If I understand your comment correctly, if newPod.Spec.Containers adds / removes / reorders containers, but also has container resource updates, we should expect that to fail validation (and not apply the container resource updates)?

---

In addition to what you said about the loop just overwriting newPod.Spec.Containers before validation (so any reordering/renaming could not possibly be caught by validation anyways), I also found it odd that `dropNonResizeUpdates` has [this comment](https://github.com/kubernetes/kubernetes/blob/052d7a5310b0bab58f81519253e08f456f39e8f3/pkg/registry/core/pod/strategy.go#L298) saying that "it will get caught in validation", but in the [actual validation function](https://github.com/kubernetes/kubernetes/blob/052d7a5310b0bab58f81519253e08f456f39e8f3/pkg/apis/core/validation/validation.go#L5581) there is no such check that occurs.

I'm planning to update PR with the following fixes to address this:
- `dropNonResizeUpdates` should just return the new podspec as-is if it detects attempts to add / remove / rename / reorder containers, so that the new podspec errors can be caught by the validation
- add the validation to `ValidatePodResize`
- add corresponding unit tests for `dropNonResizeUpdates` and `ValidatePodResize` for adding / removing / renaming / reodering
- add an e2e test for reordering to confirm that the validation error actually happens